include/networkit/graph/Graph.hpp

```diff
@@ -419,11 +422,352 @@ class Graph final {
+            return NodeIterator(G, G->upperNodeIdBound());
+        }
+    };
+
+    // Necessary for friendship with EdgeIterator.
+    class EdgeWeightIterator;
+
+    /**
+     * Class to iterate over the edges of the graph. If the graph is undirected, operator*() returns
+     * the edges (u, v) s.t. u <= v.
+     */
+    class EdgeIterator {
+        friend class EdgeWeightIterator;
+
+        const Graph *G;
+        NodeIterator nodeIter;
+        index i;
+
+        // Whether the current edge exists, or the iteration finished.
+        bool inValidState() const noexcept {
+            return nodeIter == G->nodeRange().end()
+                   || (G->degree(*nodeIter)
+                       && (G->isDirected() || (*nodeIter < G->getIthNeighbor(*nodeIter, i))));
+        }
+
+        void nextEdge() {
+            do {
+                if (++i >= G->degree(*nodeIter)) {
+                    ++nodeIter;
+                    i = 0;
+                }
+            } while (!inValidState());
+        }
+
+        void prevEdge() {
+            do {
+                if (!i) {
+                    --nodeIter;
+                    i = (nodeIter == G->nodeRange().end()
+                             ? 0
+                             : (G->degree(*nodeIter) ? G->degree(*nodeIter) - 1 : 0));
+                } else {
+                    --i;
+                }
+            } while (!inValidState());
+        }
+
+    public:
+        // The value type of the edges (i.e. a pair). Returned by operator*().
+        using value_type = Edge;
+
+        // Reference to the value_type, required by STL.
+        using reference = value_type &;
+
+        // Pointer to the value_type, required by STL.
+        using pointer = value_type *;
+
+        // STL iterator category.
+        using iterator_category = std::forward_iterator_tag;
+
+        // Signed integer type of the result of subtracting two pointers,
+        // required by STL.
+        using difference_type = ptrdiff_t;
+
+        // Own type.
+        using self = EdgeIterator;
+
+        EdgeIterator(const Graph *G, NodeIterator nodeIter)
+            : G(G), nodeIter(nodeIter), i(index{0}) {
+            if (!inValidState()) {
+                nextEdge();
+            }
+        }
+
+        /**
+         * @brief WARNING: This constructor is required for Python and should not be used as the
+         * iterator is not initialized.
+         */
+        EdgeIterator() : G(nullptr) {}
+
+        virtual ~EdgeIterator() = default;
+
+        bool operator==(const EdgeIterator &rhs) const noexcept {
+            return nodeIter == rhs.nodeIter && i == rhs.i;
+        }
+
+        bool operator!=(const EdgeIterator &rhs) const noexcept { return !(*this == rhs); }
+
+        Edge operator*() const noexcept {
+            return nodeIter == G->nodeRange().end()
+                       ? Edge(none, none)
+                       : Edge(*nodeIter, G->getIthNeighbor(*nodeIter, i));
```

Comment:
There is no need to guard against a dangling iterator state. Use an assertion instead.